fix(tool_tip): stop propagation on escape key down

weronikaolejniczak opened the pull request:
## Summary

When we have `EuiIconTip` (consequently, `EuiTooltip`) within `EuiFlyout` and with the tooltip open we click "Escape" key, the flyout gets closed. The reason is, the escape keydown event is propagated from the tooltip and captured by the flyout as its parent. Simple `event.stopPropagation()` inside `EuiTooltip` seemed to do the trick.

closes #8130

## QA

**Test path:**

1. Run Storybook locally: `yarn storybook`.
2. Find the component `EuiFlyout`.
3. Close the flyout with the escape key.
4. Change the `EuiFlyoutBody` to contain a `EuiToolTip` as children.
5. Focus the tooltip and close it with the Escape key. The flyout should not close, the tooltip - yes.
6. With the tooltip closed, click Escape key again to close the flyout.

https://github.com/user-attachments/assets/34b833bd-9c62-4f9c-9810-844222f6d193

### Manual QA of other focus-trap elements

#### EuiModal

https://github.com/user-attachments/assets/b968c2b4-d219-41a8-9116-b04daa356400

#### EuiPopover

https://github.com/user-attachments/assets/8b1be95d-13de-4677-b403-e1820e6fc1da

#### EuiDataGrid fullscreen

https://github.com/user-attachments/assets/50d30f2d-c61e-4211-a30b-c843f27ef0a0

### General checklist

- Browser QA
    - [ ] ~Checked in both **light and dark** modes~ (not applicable)
    - [ ] ~Checked in **mobile**~ (not applicable)
    - [x] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**
    - [x] Checked for **accessibility** including keyboard-only and screenreader modes
- Docs site QA
    - [ ] ~Added **[documentation](https://github.com/elastic/eui/blob/main/wiki/contributing-to-eui/documenting)**~ (not applicable)
    - [ ] ~Props have proper **autodocs** (using `@default` if default values are missing) and **[playground toggles](https://github.com/elastic/eui/blob/main/wiki/contributing-to-eui/documenting/playgrounds.md)**~ (not applicable)
    - [ ] ~Checked **[Code Sandbox](https://codesandbox.io/)** works for any docs examples~ (not applicable)
- Code quality checklist
    - [x] Added or updated **[jest](https://github.com/elastic/eui/blob/main/wiki/contributing-to-eui/testing/unit-testing.md) and [cypress](https://github.com/elastic/eui/blob/main/wiki/contributing-to-eui/testing/cypress-testing.md) tests**
    - [ ] ~Updated **[visual regression tests](https://github.com/elastic/eui/blob/main/wiki/contributing-to-eui/testing/visual-regression-testing.md)**~ (not applicable)
- Release checklist
    - [x] A **[changelog](https://github.com/elastic/eui/blob/main/wiki/contributing-to-eui/documenting/changelogs.md)** entry exists and is marked appropriately.
    - [ ] ~If applicable, added the **breaking change** issue label (and filled out the breaking change checklist)~
- Designer checklist
  - [ ] ~If applicable, [file an issue](https://github.com/elastic/platform-ux-team/issues/new/choose) to update [EUI's Figma library](https://www.figma.com/community/file/964536385682658129) with any corresponding UI changes. _(This is an internal repo, if you are external to Elastic, ask a maintainer to submit this request)_~

kibanamachine commented:
Preview staging links for this PR:
- Docs site: https://eui.elastic.co/pr_8140/
- Storybook: https://eui.elastic.co/pr_8140/storybook
- New docs: https://eui.elastic.co/pr_8140/new-docs

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3054)
* Commit: c09d2b04a41ce1ad0208de96e2993d0a1c7c985b

### Failed CI Steps
* [:hammer: EUI pull request test](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3054#01932b5d-d1da-4598-beda-568aa84d03a9)

<!--buildkite-pr-comment-eui-pull-request-test-and-deploy
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3054","number":3054,"commit":"c09d2b04a41ce1ad0208de96e2993d0a1c7c985b"}],"number":3054}
buildkite-pr-comment-->

kibanamachine commented:
Preview staging links for this PR:
- Docs site: https://eui.elastic.co/pr_8140/
- Storybook: https://eui.elastic.co/pr_8140/storybook
- New docs: https://eui.elastic.co/pr_8140/new-docs

kibanamachine commented:
Preview staging links for this PR:
- Docs site: https://eui.elastic.co/pr_8140/
- Storybook: https://eui.elastic.co/pr_8140/storybook
- New docs: https://eui.elastic.co/pr_8140/new-docs

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3072)
* Commit: a6e044f171051a536a82576dd3e7184804a13fbc

### Failed CI Steps
* [:hammer: EUI pull request test](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3072#019343c9-a5be-41ed-9054-2a77ad24f5bc)

### History
* :broken_heart: [Build #3066](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3066) failed e2b25bdf222478353733656ed230b035b4402571
* :broken_heart: [Build #3054](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3054) failed c09d2b04a41ce1ad0208de96e2993d0a1c7c985b

<!--buildkite-pr-comment-eui-pull-request-test-and-deploy
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3072","number":3072,"commit":"a6e044f171051a536a82576dd3e7184804a13fbc"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3066","number":3066,"commit":"e2b25bdf222478353733656ed230b035b4402571"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3054","number":3054,"commit":"c09d2b04a41ce1ad0208de96e2993d0a1c7c985b"}],"number":3072}
buildkite-pr-comment-->

kibanamachine commented:
Preview staging links for this PR:
- Docs site: https://eui.elastic.co/pr_8140/
- Storybook: https://eui.elastic.co/pr_8140/storybook
- New docs: https://eui.elastic.co/pr_8140/new-docs

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3085)
* Commit: 43357c2824c7cb6c7f924eac8ae18065084406cc

### Failed CI Steps
* [:hammer: EUI pull request test](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3085#01934a4b-8436-462c-8270-d5933bf50598)

### History
* :broken_heart: [Build #3072](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3072) failed a6e044f171051a536a82576dd3e7184804a13fbc
* :broken_heart: [Build #3066](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3066) failed e2b25bdf222478353733656ed230b035b4402571
* :broken_heart: [Build #3054](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3054) failed c09d2b04a41ce1ad0208de96e2993d0a1c7c985b

<!--buildkite-pr-comment-eui-pull-request-test-and-deploy
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3085","number":3085,"commit":"43357c2824c7cb6c7f924eac8ae18065084406cc"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3072","number":3072,"commit":"a6e044f171051a536a82576dd3e7184804a13fbc"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3066","number":3066,"commit":"e2b25bdf222478353733656ed230b035b4402571"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3054","number":3054,"commit":"c09d2b04a41ce1ad0208de96e2993d0a1c7c985b"}],"number":3085}
buildkite-pr-comment-->

kibanamachine commented:
Preview staging links for this PR:
- Docs site: https://eui.elastic.co/pr_8140/
- Storybook: https://eui.elastic.co/pr_8140/storybook
- New docs: https://eui.elastic.co/pr_8140/new-docs

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3091)
* Commit: a4d77c52eab97478d7f847732a30d61669073ea7

### Failed CI Steps
* [:hammer: EUI pull request test](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3091#01934e81-7683-4f3b-a1bd-64fea87c6667)

### History
* :broken_heart: [Build #3085](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3085) failed 43357c2824c7cb6c7f924eac8ae18065084406cc
* :broken_heart: [Build #3072](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3072) failed a6e044f171051a536a82576dd3e7184804a13fbc
* :broken_heart: [Build #3066](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3066) failed e2b25bdf222478353733656ed230b035b4402571
* :broken_heart: [Build #3054](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3054) failed c09d2b04a41ce1ad0208de96e2993d0a1c7c985b

<!--buildkite-pr-comment-eui-pull-request-test-and-deploy
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3091","number":3091,"commit":"a4d77c52eab97478d7f847732a30d61669073ea7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3085","number":3085,"commit":"43357c2824c7cb6c7f924eac8ae18065084406cc"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3072","number":3072,"commit":"a6e044f171051a536a82576dd3e7184804a13fbc"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3066","number":3066,"commit":"e2b25bdf222478353733656ed230b035b4402571"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3054","number":3054,"commit":"c09d2b04a41ce1ad0208de96e2993d0a1c7c985b"}],"number":3091}
buildkite-pr-comment-->

kibanamachine commented:
Preview staging links for this PR:
- Docs site: https://eui.elastic.co/pr_8140/
- Storybook: https://eui.elastic.co/pr_8140/storybook
- New docs: https://eui.elastic.co/pr_8140/new-docs

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3100)
* Commit: 3bcd2b572b7be738557ee80b36585c38e5df4812

### Failed CI Steps
* [:hammer: EUI pull request test](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3100#01935347-7801-4e27-bd74-df57f42975f3)

### History
* :broken_heart: [Build #3091](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3091) failed a4d77c52eab97478d7f847732a30d61669073ea7
* :broken_heart: [Build #3085](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3085) failed 43357c2824c7cb6c7f924eac8ae18065084406cc
* :broken_heart: [Build #3072](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3072) failed a6e044f171051a536a82576dd3e7184804a13fbc
* :broken_heart: [Build #3066](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3066) failed e2b25bdf222478353733656ed230b035b4402571
* :broken_heart: [Build #3054](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3054) failed c09d2b04a41ce1ad0208de96e2993d0a1c7c985b

<!--buildkite-pr-comment-eui-pull-request-test-and-deploy
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3100","number":3100,"commit":"3bcd2b572b7be738557ee80b36585c38e5df4812"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3091","number":3091,"commit":"a4d77c52eab97478d7f847732a30d61669073ea7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3085","number":3085,"commit":"43357c2824c7cb6c7f924eac8ae18065084406cc"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3072","number":3072,"commit":"a6e044f171051a536a82576dd3e7184804a13fbc"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3066","number":3066,"commit":"e2b25bdf222478353733656ed230b035b4402571"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3054","number":3054,"commit":"c09d2b04a41ce1ad0208de96e2993d0a1c7c985b"}],"number":3100}
buildkite-pr-comment-->

kibanamachine commented:
Preview staging links for this PR:
- Docs site: https://eui.elastic.co/pr_8140/
- Storybook: https://eui.elastic.co/pr_8140/storybook
- New docs: https://eui.elastic.co/pr_8140/new-docs

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3110)
* Commit: ba8708dc7ecba903931d35bbf641e330f8fbb3e3

### History
* :broken_heart: [Build #3100](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3100) failed 3bcd2b572b7be738557ee80b36585c38e5df4812
* :broken_heart: [Build #3091](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3091) failed a4d77c52eab97478d7f847732a30d61669073ea7
* :broken_heart: [Build #3085](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3085) failed 43357c2824c7cb6c7f924eac8ae18065084406cc
* :broken_heart: [Build #3072](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3072) failed a6e044f171051a536a82576dd3e7184804a13fbc
* :broken_heart: [Build #3066](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3066) failed e2b25bdf222478353733656ed230b035b4402571
* :broken_heart: [Build #3054](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3054) failed c09d2b04a41ce1ad0208de96e2993d0a1c7c985b

<!--buildkite-pr-comment-eui-pull-request-test-and-deploy
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3110","number":3110,"commit":"ba8708dc7ecba903931d35bbf641e330f8fbb3e3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3100","number":3100,"commit":"3bcd2b572b7be738557ee80b36585c38e5df4812"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3091","number":3091,"commit":"a4d77c52eab97478d7f847732a30d61669073ea7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3085","number":3085,"commit":"43357c2824c7cb6c7f924eac8ae18065084406cc"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3072","number":3072,"commit":"a6e044f171051a536a82576dd3e7184804a13fbc"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3066","number":3066,"commit":"e2b25bdf222478353733656ed230b035b4402571"}],"number":3110}
buildkite-pr-comment-->

cee-chen reviewed (commented):
This LGTM (awesome work with the E2E test!) - but I'll defer to @mgadewoll for any final comments since I believe she solved a similar bug/issue recently. My only other thought is we might also want to either QA or write a test for the following other focus trap components:

- modals
- popovers
- various fullscreen modes (EuiDataGrid fullscreen, EuiImage fullscreen, EuiCodeBlock fullscreen(?) - not sure that last one can even have a tooltip)

weronikaolejniczak commented:
Thanks for the great suggestion, Cee! 🌈  I focused so much on fixing that one test case I forgot about the other overlay elements that might suffer the same issue.

I updated the description with the screen recordings from the manual QA. `EuiImage` and `EuiCodeBlock` don't seem to allow for additional children in the fullscreen mode.

One doubt I have - is the `stopPropagation` in the tooltip the only point-of-failure for all these cases? If so, is it okay that we're adding (rather) expensive Cypress tests for each affected component? We could switch to RTL but then we need to remember it works with a dumb-downed DOM implementation, maybe not the best fit for asserting keyboard navigation and overlay elements.

If we agree to add the automated tests after all, I'll work on that first thing Monday!

kibanamachine commented:
Preview staging links for this PR:
- Docs site: https://eui.elastic.co/pr_8140/
- Storybook: https://eui.elastic.co/pr_8140/storybook
- New docs: https://eui.elastic.co/pr_8140/new-docs

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3123)
* Commit: 7acdd8b3f2aac6b5877e7d1f5ced39d602af0e8c

### History
* :green_heart: [Build #3110](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3110) succeeded ba8708dc7ecba903931d35bbf641e330f8fbb3e3
* :broken_heart: [Build #3100](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3100) failed 3bcd2b572b7be738557ee80b36585c38e5df4812
* :broken_heart: [Build #3091](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3091) failed a4d77c52eab97478d7f847732a30d61669073ea7
* :broken_heart: [Build #3085](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3085) failed 43357c2824c7cb6c7f924eac8ae18065084406cc
* :broken_heart: [Build #3072](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3072) failed a6e044f171051a536a82576dd3e7184804a13fbc
* :broken_heart: [Build #3066](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3066) failed e2b25bdf222478353733656ed230b035b4402571

<!--buildkite-pr-comment-eui-pull-request-test-and-deploy
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3123","number":3123,"commit":"7acdd8b3f2aac6b5877e7d1f5ced39d602af0e8c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3110","number":3110,"commit":"ba8708dc7ecba903931d35bbf641e330f8fbb3e3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3100","number":3100,"commit":"3bcd2b572b7be738557ee80b36585c38e5df4812"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3091","number":3091,"commit":"a4d77c52eab97478d7f847732a30d61669073ea7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3085","number":3085,"commit":"43357c2824c7cb6c7f924eac8ae18065084406cc"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3072","number":3072,"commit":"a6e044f171051a536a82576dd3e7184804a13fbc"}],"number":3123}
buildkite-pr-comment-->

mgadewoll commented on packages/eui/src/components/tool_tip/tool_tip.tsx:
```diff
@@ -280,6 +280,7 @@ export class EuiToolTip extends Component<EuiToolTipProps, State> {
 
   onEscapeKey = (event: React.KeyboardEvent<HTMLSpanElement>) => {
     if (event.key === keys.ESCAPE) {
+      event.stopPropagation();
```
One thing: This means the `Escape` key will never propagate. We might rather want to stop propagation only conditionally when it's visible.
If the tooltip is closed it would be expected that `Escape` closes the Flyout, otherwise users would first need to move off the tooltip anchor to be able to close it or would even be stuck in wrapper elements where there is no other element to navigate too. 🤔 

_static - blocking_

https://github.com/user-attachments/assets/c535fd77-c5d0-44c9-8d65-5e10002c68bb

_conditional_

https://github.com/user-attachments/assets/804551b0-3751-4df4-b7d3-54c9e2c4940e

weronikaolejniczak replied:
@mgadewoll that's true, the way I understood it is if the tooltip trigger is focused, the Escape shouldn't close the wrapping element. Is it something we can consult with an accessibility contractor? Or maybe it's something obvious that I'm not aware about 😅

mgadewoll replied:
While the tooltip is open, I'd agree that it should not close any wrapping element. But while it's closed I'd think it's more useful to free up the `Escape` key and prevent blocking. 

I would see the behavior as follows:

- trigger is focused -> tooltip opens (active tooltip)
- `Escape` is pressed -> tooltip closes (inactive tooltip) -> trigger stays focused as that's where the DOM focus was all the time already (no new focus on trigger)
- `Enter` is pressed -> opens tooltip (active tooltip)
OR
- `Escape` is pressed -> closes containing element (e.g. flyout or similar)

@alexwizp Do you have any insights here?

weronikaolejniczak replied:
@cee-chen @mgadewoll I conditionally call `stopPropagation` (I leveraged the `this.state.visible` bool), I moved the tests to the `EuiToolTip` suite and added `EuiModal` and `EuiPopover` cases, and I re-tested manually all the cases mentioned before by @cee-chen and updated the description with the recordings.

I'd appreciate a re-review 🙏🏻 and thanks again for the awesome suggestions!

kibanamachine commented:
Preview staging links for this PR:
- Docs site: https://eui.elastic.co/pr_8140/
- Storybook: https://eui.elastic.co/pr_8140/storybook
- New docs: https://eui.elastic.co/pr_8140/new-docs

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3134)
* Commit: 61d6291988a3d32e700f3b7c54fbf283dd23c356

### History
* :green_heart: [Build #3123](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3123) succeeded 7acdd8b3f2aac6b5877e7d1f5ced39d602af0e8c
* :green_heart: [Build #3110](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3110) succeeded ba8708dc7ecba903931d35bbf641e330f8fbb3e3
* :broken_heart: [Build #3100](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3100) failed 3bcd2b572b7be738557ee80b36585c38e5df4812
* :broken_heart: [Build #3091](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3091) failed a4d77c52eab97478d7f847732a30d61669073ea7
* :broken_heart: [Build #3085](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3085) failed 43357c2824c7cb6c7f924eac8ae18065084406cc
* :broken_heart: [Build #3072](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3072) failed a6e044f171051a536a82576dd3e7184804a13fbc

<!--buildkite-pr-comment-eui-pull-request-test-and-deploy
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3134","number":3134,"commit":"61d6291988a3d32e700f3b7c54fbf283dd23c356"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3123","number":3123,"commit":"7acdd8b3f2aac6b5877e7d1f5ced39d602af0e8c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3110","number":3110,"commit":"ba8708dc7ecba903931d35bbf641e330f8fbb3e3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3100","number":3100,"commit":"3bcd2b572b7be738557ee80b36585c38e5df4812"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3091","number":3091,"commit":"a4d77c52eab97478d7f847732a30d61669073ea7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3085","number":3085,"commit":"43357c2824c7cb6c7f924eac8ae18065084406cc"}],"number":3134}
buildkite-pr-comment-->

weronikaolejniczak commented:
@cee-chen @mgadewoll I conditionally call `stopPropagation` (I leveraged the `this.state.visible` bool), I moved the tests to the `EuiToolTip` suite and added `EuiModal` and `EuiPopover` cases, and I re-tested manually all the cases mentioned before by @cee-chen and updated the description with the recordings.

I'd appreciate a re-review 🙏🏻 and thanks again for the awesome suggestions!

kibanamachine commented:
Preview staging links for this PR:
- Docs site: https://eui.elastic.co/pr_8140/
- Storybook: https://eui.elastic.co/pr_8140/storybook
- New docs: https://eui.elastic.co/pr_8140/new-docs

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3145)
* Commit: 56ad6edf420a8c90ed81c5b8015fa8ac61c621ae

### Failed CI Steps
* [:hammer: EUI pull request test](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3145#01936d5e-ec8e-491c-b46a-4eb716fbddc0)

### History
* :green_heart: [Build #3134](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3134) succeeded 61d6291988a3d32e700f3b7c54fbf283dd23c356
* :green_heart: [Build #3123](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3123) succeeded 7acdd8b3f2aac6b5877e7d1f5ced39d602af0e8c
* :green_heart: [Build #3110](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3110) succeeded ba8708dc7ecba903931d35bbf641e330f8fbb3e3
* :broken_heart: [Build #3100](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3100) failed 3bcd2b572b7be738557ee80b36585c38e5df4812
* :broken_heart: [Build #3091](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3091) failed a4d77c52eab97478d7f847732a30d61669073ea7
* :broken_heart: [Build #3085](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3085) failed 43357c2824c7cb6c7f924eac8ae18065084406cc

<!--buildkite-pr-comment-eui-pull-request-test-and-deploy
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3145","number":3145,"commit":"56ad6edf420a8c90ed81c5b8015fa8ac61c621ae"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3134","number":3134,"commit":"61d6291988a3d32e700f3b7c54fbf283dd23c356"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3123","number":3123,"commit":"7acdd8b3f2aac6b5877e7d1f5ced39d602af0e8c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3110","number":3110,"commit":"ba8708dc7ecba903931d35bbf641e330f8fbb3e3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3100","number":3100,"commit":"3bcd2b572b7be738557ee80b36585c38e5df4812"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3091","number":3091,"commit":"a4d77c52eab97478d7f847732a30d61669073ea7"}],"number":3145}
buildkite-pr-comment-->

cee-chen commented:
@weronikaolejniczak It looks like the added Cypress test is failing on React 17/16 which is odd 👀 You may want to `yarn test-cypress-dev --react-version=17` to debug what's going on

weronikaolejniczak commented:
@cee-chen yes, I saw, I'm debugging with React 16. I'll ping you once I fix the tests.

weronikaolejniczak commented:
@cee-chen the solution was to replace `realMount` with `mount`.

**Context:**
I’m not exactly sure why but two tests for `EuiFlyout` and `EuiModal` were failing on React 16 and 17 because they couldn’t find the tooltip element. All the tests were using `realMount` which sets focus on the testing window. Something must have changed between versions, leading to different focus behavior. In earlier versions, tabbing didn't reach the tooltip trigger so the tooltip didn't show.

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3152)
* Commit: a696364c3bf9e7b5bfda5447ec1992df442bce54

### Failed CI Steps
* [:hammer: EUI pull request test](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3152#01937237-af7b-4bcb-9804-0e86ecf4af1c)

### History
* :broken_heart: [Build #3145](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3145) failed 56ad6edf420a8c90ed81c5b8015fa8ac61c621ae
* :green_heart: [Build #3134](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3134) succeeded 61d6291988a3d32e700f3b7c54fbf283dd23c356
* :green_heart: [Build #3123](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3123) succeeded 7acdd8b3f2aac6b5877e7d1f5ced39d602af0e8c
* :green_heart: [Build #3110](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3110) succeeded ba8708dc7ecba903931d35bbf641e330f8fbb3e3
* :broken_heart: [Build #3100](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3100) failed 3bcd2b572b7be738557ee80b36585c38e5df4812

<!--buildkite-pr-comment-eui-pull-request-test-and-deploy
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3152","number":3152,"commit":"a696364c3bf9e7b5bfda5447ec1992df442bce54"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3145","number":3145,"commit":"56ad6edf420a8c90ed81c5b8015fa8ac61c621ae"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3134","number":3134,"commit":"61d6291988a3d32e700f3b7c54fbf283dd23c356"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3123","number":3123,"commit":"7acdd8b3f2aac6b5877e7d1f5ced39d602af0e8c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3110","number":3110,"commit":"ba8708dc7ecba903931d35bbf641e330f8fbb3e3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3100","number":3100,"commit":"3bcd2b572b7be738557ee80b36585c38e5df4812"}],"number":3152}
buildkite-pr-comment-->

kibanamachine commented:
Preview staging links for this PR:
- Docs site: https://eui.elastic.co/pr_8140/
- Storybook: https://eui.elastic.co/pr_8140/storybook
- New docs: https://eui.elastic.co/pr_8140/new-docs

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3154)
* Commit: 6dc6d35f68dd131df96ed196a11132d246849da0

### History
* :broken_heart: [Build #3152](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3152) failed a696364c3bf9e7b5bfda5447ec1992df442bce54
* :broken_heart: [Build #3145](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3145) failed 56ad6edf420a8c90ed81c5b8015fa8ac61c621ae
* :green_heart: [Build #3134](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3134) succeeded 61d6291988a3d32e700f3b7c54fbf283dd23c356
* :green_heart: [Build #3123](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3123) succeeded 7acdd8b3f2aac6b5877e7d1f5ced39d602af0e8c
* :green_heart: [Build #3110](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3110) succeeded ba8708dc7ecba903931d35bbf641e330f8fbb3e3
* :broken_heart: [Build #3100](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3100) failed 3bcd2b572b7be738557ee80b36585c38e5df4812

<!--buildkite-pr-comment-eui-pull-request-test-and-deploy
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3154","number":3154,"commit":"6dc6d35f68dd131df96ed196a11132d246849da0"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3152","number":3152,"commit":"a696364c3bf9e7b5bfda5447ec1992df442bce54"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3145","number":3145,"commit":"56ad6edf420a8c90ed81c5b8015fa8ac61c621ae"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3134","number":3134,"commit":"61d6291988a3d32e700f3b7c54fbf283dd23c356"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3123","number":3123,"commit":"7acdd8b3f2aac6b5877e7d1f5ced39d602af0e8c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3110","number":3110,"commit":"ba8708dc7ecba903931d35bbf641e330f8fbb3e3"}],"number":3154}
buildkite-pr-comment-->

mgadewoll commented:
> @cee-chen I’m not exactly sure why but two tests for `EuiFlyout` and `EuiModal` were failing on React 16 and 17 because they couldn’t find the tooltip element. I tried switching from `realMount` to `mount` which would make the tests for the old versions pass but for 18 fail. Something must have changed between versions, leading to different focus behavior.
> 
> The solution was to focus the tool-tip trigger directly instead of tabbing into it to assure consistent behavior between versions. Otherwise, the tabbing would start and end at different elements depending on the version.


@weronikaolejniczak For testing with key presses, I generally would prefer to be able to use proper `realPress()` to move to the tested component instead of direct `focus()` to ensure all keyboard functionality/flow works as expected.
From what I see, the issue is that `realMount` adds a click on mount to move focus to the DOM.
Since we're testing components here that have default click-away behavior, the tests fail because the components close because of the click.
We could enhance the `realMount` to accept `options` to define the initial click position to manually adjust it where needed BUT that would likely become annoying for positioning it properly.

Instead we could still use `mount()` but use `focus()` to programmatically define the start position on the container element (considering other focus to that element is tested in the component specific tests anyway) and proceed with `realPress()`?

```tsx
// example for Flyout
cy.mount(
  <Flyout>
    <EuiToolTip content="Tooltip text here" data-test-subj="tool_tip">
      <EuiButton data-test-subj="tool_tip_trigger">Show tooltip</EuiButton>
    </EuiToolTip>
  </Flyout>
);
cy.get('[data-test-subj="flyout"]').focus(); // move focus to flyout
cy.get('[data-test-subj="tool_tip"]').should('not.exist');

cy.repeatRealPress('Tab', 2);
cy.get('[data-test-subj="tool_tip"]').should('exist');

cy.realPress('Escape');
cy.get('[data-test-subj="tool_tip"]').should('not.exist');
cy.get('[data-test-subj="flyout"]').should('exist');

cy.realPress('Escape');
cy.get('[data-test-subj="flyout"]').should('not.exist');
```

weronikaolejniczak commented:
@mgadewoll that's a fantastic suggestion, thank you, Lene! 🙌🏻 I pushed the changes for all test cases, could you check?

kibanamachine commented:
Preview staging links for this PR:
- Docs site: https://eui.elastic.co/pr_8140/
- Storybook: https://eui.elastic.co/pr_8140/storybook
- New docs: https://eui.elastic.co/pr_8140/new-docs

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3160)
* Commit: f020bb0a0f38aac49d201ee1e4b2d246a0f9f824

### History
* :green_heart: [Build #3154](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3154) succeeded 6dc6d35f68dd131df96ed196a11132d246849da0
* :broken_heart: [Build #3152](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3152) failed a696364c3bf9e7b5bfda5447ec1992df442bce54
* :broken_heart: [Build #3145](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3145) failed 56ad6edf420a8c90ed81c5b8015fa8ac61c621ae
* :green_heart: [Build #3134](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3134) succeeded 61d6291988a3d32e700f3b7c54fbf283dd23c356
* :green_heart: [Build #3123](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3123) succeeded 7acdd8b3f2aac6b5877e7d1f5ced39d602af0e8c
* :green_heart: [Build #3110](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3110) succeeded ba8708dc7ecba903931d35bbf641e330f8fbb3e3

<!--buildkite-pr-comment-eui-pull-request-test-and-deploy
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3160","number":3160,"commit":"f020bb0a0f38aac49d201ee1e4b2d246a0f9f824"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3154","number":3154,"commit":"6dc6d35f68dd131df96ed196a11132d246849da0"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3152","number":3152,"commit":"a696364c3bf9e7b5bfda5447ec1992df442bce54"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3145","number":3145,"commit":"56ad6edf420a8c90ed81c5b8015fa8ac61c621ae"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3134","number":3134,"commit":"61d6291988a3d32e700f3b7c54fbf283dd23c356"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/3123","number":3123,"commit":"7acdd8b3f2aac6b5877e7d1f5ced39d602af0e8c"}],"number":3160}
buildkite-pr-comment-->
